add basic support of references to POST/GET subjects

libretto opened the pull request:
<!-- All contributors please complete these sections, including maintainers -->
# About this change - What it does

<!-- Provide a small sentence that summarizes the change. -->

<!-- Provide the issue number below if it exists. -->
References: #xxxxx

# Why this way

<!-- Provide a small explanation on why this is the approach you took for solving this problem. -->

jjaakola-aiven commented on karapace/schema_registry_apis.py:
```diff
@@ -639,6 +703,52 @@ async def subject_version_schema_get(self, content_type, *, subject, version):
+                status=HTTPStatus.UNPROCESSABLE_ENTITY,
+            )
+
+        referenced_by = self.ksr.referenced_by.get(str(subject) + "_" + str(version), None)
+        if not referenced_by:
+            referenced_by = list()
```
```suggestion
        referenced_by = self.ksr.referenced_by.get(str(subject) + "_" + str(version), [])
```

jjaakola-aiven commented on karapace/schema_registry_apis.py:
```diff
@@ -781,8 +924,37 @@ def write_new_schema_local(
+        new_schema_references = None
+        references = body.get("references")
+        if references:
+            if schema_type != SchemaType.PROTOBUF:
```
This should be validated before calling `write_new_schema_local`.

jjaakola-aiven commented on tests/integration/test_client.py:
```diff
@@ -9,7 +9,7 @@ async def test_remote_client(registry_async_client: Client) -> None:
     reg_cli.client = registry_async_client
     subject = new_random_name("subject")
-    sc_id = await reg_cli.post_new_schema(subject, schema_avro)
+    sc_id = await reg_cli.post_new_schema(subject, schema_avro)  # pylint: disable=E1120
```
Is the `pylint` disable necessary? Same few lines below.

jjaakola-aiven commented on tests/integration/test_client_protobuf.py:
```diff
@@ -25,10 +25,10 @@ async def test_remote_client_protobuf2(registry_async_client):
     stored_schema = await reg_cli.get_schema_for_id(sc_id)
     assert stored_schema == schema_protobuf, f"stored schema {stored_schema} is not {schema_protobuf}"
-    stored_id, stored_schema = await reg_cli.get_latest_schema(subject)
+    stored_id, stored_schema = await reg_cli.get_latest_schema(subject)  # pylint: disable=unused-variable
```
Is the pylint disable necessary, both returned variables are used?

jjaakola-aiven reviewed (changes requested):
Sorry for the delay on reviewing this.
I've done first pass and some comments. I have not looked the testing coverage yet.

Compatibility check API does not support references, `/compatibility/subjects/<subject:path>/versions/<version:path>`. I am not sure how CSR uses it, but worth checking.

sujayinstaclustr commented:
@jjaakola-aiven We address all the review feedback. Let us know if we can raise a PR against the upstream.

libretto commented on karapace/schema_models.py:
```diff
@@ -146,9 +177,32 @@ def parse(schema_type: SchemaType, schema_str: str) -> "ValidatedTypedSchema":
+
+
+class References:
+    def __init__(self, schema_type: SchemaType, references: JsonData):
```
> I would propose to type the references to specific type instead of `JsonData`.

jjaakola-aiven replied:
Probably just an error from me on review, not sure after vacation. Possibly I've been thinking on modeling the data directly in the controller layer from JSON to objects.

libretto commented on karapace/schema_models.py:
```diff
@@ -84,6 +89,7 @@ def __init__(self, schema_type: SchemaType, schema_str: str):
         """
         self.schema_type = schema_type
         self.schema_str = schema_str
+        self.references = references
```
done

libretto commented:
> @jjaakola-aiven We addressed some of the feedback and waiting for to discuss few open comments (the ones showing pending)
Pending was finished. I missed committing the review.

sujayinstaclustr commented:
> 

@libretto there are conflicts in the code pushed. Can you resolve the conflicts?

sujayinstaclustr commented:
@jjaakola-aiven If this PR is all good to merge and we can raise a PR to upstream

jjaakola-aiven commented on karapace/schema_reader.py:
```diff
@@ -375,21 +377,36 @@ def _handle_msg_schema(self, key: dict, value: Optional[dict]) -> None:
+                    ref_str = reference_key(ref["subject"], ref["version"])
+                    referents = self.referenced_by.get(ref_str, None)
+                    if referents:
+                        LOG.info("Adding entry subject referenced_by : %r", ref_str)
```
Debug log.

libretto replied:
Must we remove it?

jjaakola-aiven replied:
If deemed valuable for debugging no need to remove, but change it to `LOG.debug`.

jjaakola-aiven commented on karapace/schema_reader.py:
```diff
@@ -375,21 +377,36 @@ def _handle_msg_schema(self, key: dict, value: Optional[dict]) -> None:
+                        LOG.info("Adding entry subject referenced_by : %r", ref_str)
+                        referents.append(schema_id)
+                    else:
+                        LOG.info("Adding entry subject referenced_by : %r", ref_str)
```
Debug log.

jjaakola-aiven commented on karapace/schema_registry_apis.py:
```diff
@@ -44,6 +45,10 @@ class SchemaErrorCodes(Enum):
     INVALID_VERSION_ID = 42202
     INVALID_COMPATIBILITY_LEVEL = 42203
     INVALID_AVRO_SCHEMA = 44201
+    INVALID_PROTOBUF_SCHEMA = 44202
+    INVALID_REFERECES = 44203
+    REFERENCES_SUPPORT_NOT_IMPLEMENTED = 44501
+    SCHEMAVERSION_HAS_REFERENCES = 44503
```
I would reserve error codes 44201 to 442xx for Avro, Protobuf, JSONSchema and future formats.

Nit: typo in `INVALID_REFERECES`.

Maybe following? 
```suggestion
    INVALID_REFERENCES = 44301
    REFERENCES_SUPPORT_NOT_IMPLEMENTED = 44302
    SCHEMAVERSION_HAS_REFERENCES = 44303
```

jjaakola-aiven commented on karapace/schema_reader.py:
```diff
@@ -366,20 +370,33 @@ def _handle_msg_schema(self, key: dict, value: Optional[dict]) -> None:
             subjects_schemas[schema_version] = schema
+            if schema_references:
+                for ref in schema_references:
+                    ref_str = str(ref["subject"]) + "_" + str(ref["version"])
```
Sure, function is good approach here.

jjaakola-aiven commented on tests/integration/test_schema_protobuf.py:
```diff
@@ -104,3 +104,55 @@ async def test_protobuf_schema_compatibility(registry_async_client: Client, trai
+
+    res = await registry_async_client.delete("subjects/customer/versions/1")
+    assert res.status_code == 404
+    match_msg = "Subject 'customer' Version 1 was not deleted because it is referenced by schemas with ids:[2]"
```
The lifecycle of schema and references would be nice to test:
1. create schema A
1. create schema B referencing A
1. try to delete A
1. delete B
1. delete A


Steps 1-3 are covered, steps 4 and 5 would be needed.

jjaakola-aiven commented on karapace/schema_registry_apis.py:
```diff
@@ -55,6 +60,7 @@ class SchemaErrorMessages(Enum):
         "forward, full, backward_transitive, forward_transitive, and "
         "full_transitive"
     )
+    REFERENCES_SUPPORT_NOT_IMPLEMENTED = "Schema references are not supported for '{schema_type}' schema type yet"
```
The `yet` is not needed and the request could contain arbitrary string as schema type resulting `Schema references are not supported for '<ARBITRARY>' schema type yet`

libretto replied:
The request always contains JSON with the exact schema type definition from the set of AVRO, JSON, and PROTOBUF in other case we raise an error.

jjaakola-aiven replied:
I meant that `yet` points to future support. I would just state in the error that not supported without hinting on possible future support.

jjaakola-aiven reviewed (changes requested):
The compatibility check API does not support references, for completeness it should be added.

jjaakola-aiven commented on karapace/schema_registry_apis.py:
```diff
@@ -736,10 +780,29 @@ async def _subject_version_delete_local(self, content_type: str, subject: str, v
+        if referenced_by and len(referenced_by) > 0:
+            self.r(
+                body={
+                    "error_code": SchemaErrorCodes.SCHEMAVERSION_HAS_REFERENCES.value,
```
I cross checked this error code and correct would be: https://github.com/confluentinc/schema-registry/blob/master/core/src/main/java/io/confluent/kafka/schemaregistry/rest/exceptions/Errors.java#L63

jjaakola-aiven commented on karapace/schema_models.py:
```diff
@@ -67,6 +68,10 @@ class InvalidSchema(Exception):
     pass
 
 
+class InvalidReferences(Exception):
```
Seems that this is not raised.
